clj-refactor.el

```diff
@@ -3992,18 +4000,25 @@ See: https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-change-function-
       (pop-to-buffer cljr--change-signature-buffer))))
 
 ;;;###autoload
-(defun cljr--inject-jack-in-dependencies ()
+(defun cljr--inject-jack-in-dependencies (cider-jack-in &rest args)
   "Inject the REPL dependencies of clj-refactor at `cider-jack-in'.
 If injecting the dependencies is not preferred set `cljr-inject-dependencies-at-jack-in' to nil."
-  (when (and cljr-inject-dependencies-at-jack-in
-             (boundp 'cider-jack-in-lein-plugins)
-             (boundp 'cider-jack-in-nrepl-middlewares))
-    (add-to-list 'cider-jack-in-lein-plugins `("refactor-nrepl" ,(cljr--version t)))
-    (add-to-list 'cider-jack-in-nrepl-middlewares "refactor-nrepl.middleware/wrap-refactor")))
+  (if (and cljr-inject-dependencies-at-jack-in
+           (boundp 'cider-jack-in-lein-plugins)
+           (boundp 'cider-jack-in-nrepl-middlewares)
+           (not (string-empty-p (cljr--project-dir))))
+      (let ((cider-jack-in-lein-plugins
+             (cons `("refactor-nrepl" ,(cljr--version t))
+                   cider-jack-in-lein-plugins))
+            (cider-jack-in-nrepl-middlewares
+             (cons "refactor-nrepl.middleware/wrap-refactor"
+                   cider-jack-in-nrepl-middlewares)))
+        (apply cider-jack-in args))
+    (apply cider-jack-in args)))
 
 ;;;###autoload
 (eval-after-load 'cider
-  '(cljr--inject-jack-in-dependencies))
+  '(advice-add #'cider-jack-in :around #'cljr--inject-jack-in-dependencies))
```

Comment:
@bbatsov This seems like a no-no to me.  What are you thoughts on using an advice hack here?

For context, @raxod502 is submitting a patch so a variable will prevent the middleware dep from even being injected when we're outside a project.  The reason being that the middleware will emit warnings that he doesn't care about.

Reply:
The reason I used advice here was that advice was already being used elsewhere; see https://github.com/clojure-emacs/clj-refactor.el/blob/f5295df68955c23fffd60718039fd386d13c77f5/clj-refactor.el#L1638.

Note also that mutating a global variable in an autoload is not considered good practice either, so I don't think either the previous code or the new code is ideal.

Reply:
If you really don't want to use advice, I'll be happy to submit a pull request upstream for some additional customizable variables to be added so that doing this doesn't require advice.